plugwise/nodes/__init__.py

```diff
@@ -1,4 +1,417 @@
+    FEATURE_RSSI_OUT,
+    PRIORITY_LOW,
+    UTF8_DECODE,
+)
+from ..messages.requests import NodeFeaturesRequest, NodeInfoRequest, NodePingRequest
+from ..messages.responses import (
+    NodeFeaturesResponse,
+    NodeInfoResponse,
+    NodeJoinAckResponse,
+    NodePingResponse,
+    NodeResponse,
+)
+from ..util import validate_mac, version_to_model
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class PlugwiseNode:
+    """Base class for a Plugwise node."""
+
+    def __init__(self, mac, address, message_sender):
+        mac = mac.upper()
+        if not validate_mac(mac):
+            _LOGGER.debug(
+                "MAC address is in unexpected format: %s",
+                str(mac),
+            )
+        self._mac = bytes(mac, encoding=UTF8_DECODE)
+        self.message_sender = message_sender
+        self._features = ()
+        self._address = address
+        self._callbacks = {}
+        self._last_update = None
+        self._available = False
+        self._battery_powered = False
+        self._measures_power = False
+        self._RSSI_in = None
+        self._RSSI_out = None
+        self._ping = None
+        self._node_type = None
+        self._hardware_version = None
+        self._firmware_version = None
+        self._relay_state = False
+        self._last_log_address = None
+        self.last_info_message = None
+        self._device_features = None
+
+    @property
+    def available(self) -> bool:
+        """Current network state of plugwise node."""
+        return self._available
+
+    @available.setter
+    def available(self, state: bool):
+        """Set current network availability state of plugwise node."""
+        if state:
+            if not self._available:
+                self._available = True
+                _LOGGER.debug(
+                    "Mark node %s available",
+                    self.mac,
+                )
+                self.do_callback(FEATURE_AVAILABLE["id"])
+        else:
+            if self._available:
+                self._available = False
+                _LOGGER.debug(
+                    "Mark node %s unavailable",
+                    self.mac,
+                )
+                self.do_callback(FEATURE_AVAILABLE["id"])
+
+    @property
+    def battery_powered(self) -> bool:
+        """Return True if node is a SED (battery powered) device."""
+        return self._battery_powered
+
+    @property
+    def hardware_model(self) -> str:
+        """Return hardware model."""
+        if self._hardware_version:
+            return version_to_model(self._hardware_version)
+        return None
+
+    @property
+    def hardware_version(self) -> str:
+        """Return hardware version."""
+        if self._hardware_version is not None:
+            return self._hardware_version
+        return "Unknown"
+
+    @property
+    def features(self) -> tuple:
+        """Return the abstracted features supported by this plugwise device."""
+        return self._features
```

Comment:
Shouldn't this be `return self._device_features`?
I see both `self._features` and `self._device_features` are defined, but only `self._device_features` is given values, via 
`_process_features_response`.

Reply:
Update: never mind, I see the `self._features` are given values in the node Classes.